File: built-in plugins dsp/mb-strings-ensemble.cpp
/**
 * MB String Ensemble
 * Category : instrument
 * Type     : strings
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Lush orchestral string section
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STRINGS_ENSEMBLE_H
#define MB_STRINGS_ENSEMBLE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStringsEnsemble : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-strings-ensemble";
    static constexpr const char* PLUGIN_NAME    = "MB String Ensemble";
    static constexpr const char* PLUGIN_TYPE    = "strings";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float warmth = 0.6f;  // range [0, 1]
    float ensemble = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbStringsEnsemble() = default;
    ~MbStringsEnsemble() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.warmth = std::clamp(params.warmth, 0f, 1f);
        params.ensemble = std::clamp(params.ensemble, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB String Ensemble
        return input;
    }
};

#endif // MB_STRINGS_ENSEMBLE_H

File: built-in plugins dsp/mb-strings-ensemble.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbStringsEnsemblePlugin: PluginDefinition = {
    id: 'mb-strings-ensemble', slug: 'mb-strings-ensemble', name: 'MB String Ensemble', category: 'instrument', type: 'strings', version: '1.0.0',
    description: 'Lush orchestral string section', author: 'Max Booster',
    oscillators: [{ type: 'sawtooth', detune: -5, gain: 0.3 }, { type: 'sawtooth', detune: 5, gain: 0.3 }, { type: 'triangle', detune: 0, gain: 0.4 }],
    envelope: { attack: 0.3, decay: 0.5, sustain: 0.8, release: 1.0 },
    parameters: [
      { id: 'warmth', name: 'Warmth', type: 'float', defaultValue: 0.6, minValue: 0, maxValue: 1, automatable: true },
      { id: 'ensemble', name: 'Ensemble', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { warmth: 0.6, ensemble: 0.5, volume: 0.8 },
  };

export default MbStringsEnsemblePlugin;
